test/core/iomgr/threadpool_test.cc

```diff
@@ -0,0 +1,187 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "src/core/lib/iomgr/executor/threadpool.h"
+
+#include "test/core/util/test_config.h"
+
+const int kSmallThreadPoolSize = 20;
+const int kLargeThreadPoolSize = 100;
+const int kThreadSmallIter = 100;
+const int kThreadLargeIter = 10000;
+
+// Simple functor for testing. It will count how many times being called.
+class SimpleFunctorForAdd : public grpc_experimental_completion_queue_functor {
+ public:
+  friend class SimpleFunctorCheckForAdd;
+  SimpleFunctorForAdd() {
+    functor_run = &SimpleFunctorForAdd::Run;
+    internal_next = this;
+    internal_success = 0;
+  }
+  ~SimpleFunctorForAdd() {}
+  static void Run(struct grpc_experimental_completion_queue_functor* cb,
+                  int ok) {
+    auto* callback = static_cast<SimpleFunctorForAdd*>(cb);
+    callback->count_.FetchAdd(1, grpc_core::MemoryOrder::RELAXED);
+  }
+
+  int count() { return count_.Load(grpc_core::MemoryOrder::RELAXED); }
+
+ private:
+  grpc_core::Atomic<int> count_{0};
+};
+
+// Checks the given SimpleFunctorForAdd's count with a given number.
+class SimpleFunctorCheckForAdd
+    : public grpc_experimental_completion_queue_functor {
+ public:
+  SimpleFunctorCheckForAdd(int ok, int* count) : count_(count) {
+    functor_run = &SimpleFunctorCheckForAdd::Run;
+    internal_success = ok;
+  }
+  ~SimpleFunctorCheckForAdd() {}
+  static void Run(struct grpc_experimental_completion_queue_functor* cb,
+                  int ok) {
+    auto* callback = static_cast<SimpleFunctorCheckForAdd*>(cb);
+    (*callback->count_)++;
+    GPR_ASSERT(*callback->count_ == callback->internal_success);
+  }
+
+ private:
+  int* count_;
+};
+
+static void test_add(void) {
+  gpr_log(GPR_INFO, "test_add");
+  grpc_core::ThreadPool* pool =
+      grpc_core::New<grpc_core::ThreadPool>(kSmallThreadPoolSize, "test_add");
+
+  SimpleFunctorForAdd* functor = grpc_core::New<SimpleFunctorForAdd>();
+  for (int i = 0; i < kThreadSmallIter; ++i) {
+    pool->Add(functor);
+  }
+  grpc_core::Delete(pool);
+  GPR_ASSERT(functor->count() == kThreadSmallIter);
+  grpc_core::Delete(functor);
+  gpr_log(GPR_DEBUG, "Done.");
+}
+
+// Thread that adds closures to pool
+class WorkThread {
+ public:
+  WorkThread(grpc_core::ThreadPool* pool, SimpleFunctorForAdd* cb, int num_add)
+      : num_add_(num_add), cb_(cb), pool_(pool) {
+    thd_ = grpc_core::Thread(
+        "thread_pool_test_add_thd",
+        [](void* th) { static_cast<WorkThread*>(th)->Run(); }, this);
+  }
+  ~WorkThread() {}
+
+  void Start() { thd_.Start(); }
+  void Join() { thd_.Join(); }
+
+ private:
+  void Run() {
+    for (int i = 0; i < num_add_; ++i) {
+      pool_->Add(cb_);
+    }
+  }
+
+  int num_add_;
+  SimpleFunctorForAdd* cb_;
+  grpc_core::ThreadPool* pool_;
+  grpc_core::Thread thd_;
+};
+
+static void test_constructor(void) {
+  // Size is 0 case
+  grpc_core::ThreadPool* pool_size_zero =
```

Comment:
Let us test different attributes (num_threads == 0 and options) in separate test cases.